examples/cvode/CXX_serial/cv_heat2D.hpp

```diff
@@ -0,0 +1,352 @@
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------------------
+ * Header file for 2D heat equation example problem using Ginkgo linear solvers.
+ * See cv_heat2D.cpp for more information.
+ * ---------------------------------------------------------------------------*/
+
+#include <cstdio>
+#include <iostream>
+#include <iomanip>
+#include <fstream>
+#include <limits>
+#include <cmath>
+#include <string>
+
+// SUNDIALS types
+#include <sundials/sundials_types.h>
+#include <sundials/sundials_nvector.h>
+
+// Common utility functions
+#include <example_utilities.hpp>
+
+// Macros for problem constants
+#define PI    RCONST(3.141592653589793238462643383279502884197169)
+#define ZERO  RCONST(0.0)
+#define ONE   RCONST(1.0)
+#define TWO   RCONST(2.0)
+#define EIGHT RCONST(8.0)
+
+// -----------------------------------------------------------------------------
+// User data structure
+// -----------------------------------------------------------------------------
+
+struct UserData
+{
+  // Diffusion coefficients in the x and y directions
+  sunrealtype kx = ONE;
+  sunrealtype ky = ONE;
+
+  // Final time
+  sunrealtype tf = ONE;
+
+  // Upper bounds in x and y directions
+  sunrealtype xu = ONE;
+  sunrealtype yu = ONE;
+
+  // Number of nodes in the x and y directions
+  sunindextype nx = 32;
+  sunindextype ny = 32;
+
+  // Total number of nodes
+  sunindextype nodes = nx * ny;
+
+  // Mesh spacing in the x and y directions
+  sunrealtype dx = xu / (nx - 1);
+  sunrealtype dy = yu / (ny - 1);
+
+  // Integrator settings
+  sunrealtype rtol     = SUN_RCONST(1.0e-4); // relative tolerance
+  sunrealtype atol     = SUN_RCONST(1.0e-8); // absolute tolerance
+  int         maxsteps = 0;                  // max number of steps between outputs
+
+  // Linear solver and preconditioner settings
+  bool     pcg      = true; // use PCG (true) or GMRES (false)
+  bool     prec     = true; // preconditioner on/off
+  int      liniters = 20;   // number of linear iterations
+  int      msbp     = 0;    // max number of steps between preconditioner setups
+  realtype epslin   = ZERO; // linear solver tolerance factor
+
+  // Inverse of Jacobian diagonal for preconditioner
+  N_Vector d = nullptr;
+
+  // Ouput variables
+  bool          output = false; // write solution to disk
+  int           nout   = 20;    // number of output times
+  std::ofstream uout;           // output file stream
+  std::ofstream eout;           // error file stream
+
+  // Destructor
+  ~UserData();
+};
+
+// Free memory allocated within Userdata
+UserData::~UserData()
+{
+  // Free preconditioner data
+  if (d)
+  {
+    N_VDestroy(d);
+    d = nullptr;
+  }
+}
+
+// -----------------------------------------------------------------------------
+// Utility functions
```

Comment:
@gardner48

Reply:
This is mostly to keep the `.cpp` files more concise and focused on the user code for setting up and running the problem. Some of the examples were getting rather long and hard to digest with the utility functions for parsing command line options and writing output. I thought this could make things cleaner but if there's a strong preference for keeping everything in the `.cpp` file that's good with me.

Reply:
I think this is a great reason to put them in the `.hpp`.